test/hardhat/irm/Irm.spec.ts

```diff
@@ -0,0 +1,95 @@
+import { AbiCoder, keccak256, toBigInt } from "ethers";
+import hre from "hardhat";
+import _range from "lodash/range";
+import { AdaptativeCurveIrm } from "types";
+import { MarketParamsStruct } from "types/lib/morpho-blue/src/interfaces/IIrm";
+
+import { SignerWithAddress } from "@nomicfoundation/hardhat-ethers/signers";
+import { setNextBlockTimestamp } from "@nomicfoundation/hardhat-network-helpers/dist/src/helpers/time";
+
+let seed = 42;
+const random = () => {
+  seed = (seed * 16807) % 2147483647;
+
+  return (seed - 1) / 2147483646;
+};
+
+const identifier = (marketParams: MarketParamsStruct) => {
+  const encodedMarket = AbiCoder.defaultAbiCoder().encode(
+    ["address", "address", "address", "address", "uint256"],
+    Object.values(marketParams),
+  );
+
+  return Buffer.from(keccak256(encodedMarket).slice(2), "hex");
+};
+
+const logProgress = (name: string, i: number, max: number) => {
+  if (i % 10 == 0) console.log("[" + name + "]", Math.floor((100 * i) / max), "%");
+};
+
+const randomForwardTimestamp = async () => {
+  const block = await hre.ethers.provider.getBlock("latest");
+  const elapsed = random() < 1 / 2 ? 0 : (1 + Math.floor(random() * 100)) * 12; // 50% of the time, don't go forward in time.
+
+  const newTimestamp = block!.timestamp + elapsed;
+
+  await setNextBlockTimestamp(block!.timestamp + elapsed);
+
+  return newTimestamp;
+};
+
+describe("irm", () => {
+  let admin: SignerWithAddress;
+
+  let irm: AdaptativeCurveIrm;
+
+  let marketParams: MarketParamsStruct;
+
+  beforeEach(async () => {
+    [admin] = await hre.ethers.getSigners();
+
+    const AdaptativeCurveIrmFactory = await hre.ethers.getContractFactory("AdaptativeCurveIrm", admin);
+
+    irm = await AdaptativeCurveIrmFactory.deploy(
+      await admin.getAddress(),
+      4000000000000000000n,
+      1585489599188n,
+      900000000000000000n,
+      317097919n,
+    );
+
+    const irmAddress = await irm.getAddress();
+
+    marketParams = {
+      // Non-zero address to include calldata gas cost.
```

Comment:
but the irm isn't called by an eoa, so it's passed via memory right?

Reply:
These params may get read from memory for the CALL from Morpho Blue to the IRM but from the IRM's POV, the calldata will be calldata and we just want to include the fact that all bytes may not be zero here

Reply:
I mean that we want to simulate that there is non-zero bytes in the CALL from Morpho to the IRM (to possibly account for the gas cost associated to it), hence why we don't only pass zero bytes

Do you mean that calldata for CALLs inside a tx (for example from Morpho to IRM) are passed as memory? In which case I am highly surprised and would like to read resources on the topic 👀

Reply:
>Do you mean that calldata for CALLs inside a tx (for example from Morpho to IRM) are passed as memory? In which case I am highly surprised and would like to read resources on the topic 👀

to me yes, let me search for that